lib/src/mock_query.dart

```diff
@@ -239,55 +251,182 @@ class MockQuery<T extends Object?> extends FakeQueryWithParent<T> {
-          if (field is String || field is FieldPath) {
-            // DocumentSnapshot.get can throw StateError
-            // if field cannot be found. In query it does not matter,
-            // so catch and set value to null.
-            try {
-              value = document.get(field);
-            } on StateError catch (_) {
-              value = null;
-            }
-          } else if (field == FieldPath.documentId) {
-            value = document.id;
-
-            // transform any DocumentReference in the query to id.
-            isEqualTo = transformValue(isEqualTo, documentReferenceToId);
-            isNotEqualTo = transformValue(isNotEqualTo, documentReferenceToId);
-            isLessThan = transformValue(isLessThan, documentReferenceToId);
-            isLessThanOrEqualTo =
-                transformValue(isLessThanOrEqualTo, documentReferenceToId);
-            isGreaterThan =
-                transformValue(isGreaterThan, documentReferenceToId);
-            isGreaterThanOrEqualTo =
-                transformValue(isGreaterThanOrEqualTo, documentReferenceToId);
-            arrayContains =
-                transformValue(arrayContains, documentReferenceToId);
-            arrayContainsAny =
-                transformValue(arrayContainsAny, documentReferenceToId);
-            whereIn = transformValue(whereIn, documentReferenceToId);
-            whereNotIn = transformValue(whereNotIn, documentReferenceToId);
-            isNull = transformValue(isNull, documentReferenceToId);
-          }
-          return _valueMatchesQuery(value,
-              isEqualTo: isEqualTo,
-              isNotEqualTo: isNotEqualTo,
-              isLessThan: isLessThan,
-              isLessThanOrEqualTo: isLessThanOrEqualTo,
-              isGreaterThan: isGreaterThan,
-              isGreaterThanOrEqualTo: isGreaterThanOrEqualTo,
-              arrayContains: arrayContains,
-              arrayContainsAny: arrayContainsAny,
-              whereIn: whereIn,
-              whereNotIn: whereNotIn,
-              isNull: isNull);
-        }).toList();
+    if (field is Filter) {
+      final predicate = _buildFilterPredicate(field.toJson());
+      return MockQuery(this, (docs) => docs.where(predicate).toList());
+    }
+
+    final operation = (List<DocumentSnapshot<T>> docs) => docs
+        .where((document) => _wherePredicate(document, field,
+            isEqualTo: isEqualTo,
+            isNotEqualTo: isNotEqualTo,
+            isLessThan: isLessThan,
+            isLessThanOrEqualTo: isLessThanOrEqualTo,
+            isGreaterThan: isGreaterThan,
+            isGreaterThanOrEqualTo: isGreaterThanOrEqualTo,
+            arrayContains: arrayContains,
+            arrayContainsAny: arrayContainsAny,
+            whereIn: whereIn,
+            whereNotIn: whereNotIn,
+            isNull: isNull))
+        .toList();
     return MockQuery<T>(this, operation);
   }
 
+  bool Function(DocumentSnapshot<T> document) _buildFilterPredicate(
```

Comment:
Very nice

Reply:
Perhaps explaining briefly in the method's docs how a Filter gets serialized to JSON could help others understand the code faster.